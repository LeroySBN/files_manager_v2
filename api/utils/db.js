// MongoDB utils
import {MongoClient} from 'mongodb';
import 'dotenv/config';

const database = process.env.DB_DATABASE || 'files_db';
const url = process.env.MONGO_URL || 'mongodb://admin:admin_password@mongodb:27017/files_db?authSource=admin';

class DBClient {
  constructor() {
    this.client = new MongoClient(url, {
      useUnifiedTopology: true,
      connectTimeoutMS: 5000,
      serverSelectionTimeoutMS: 5000,
    });

    this.connected = false;
    this.connecting = false;

    // Initial connection
    this.connect();
  }

  async connect() {
    if (this.connected || this.connecting) return;

    try {
      this.connecting = true;
      await this.client.connect();
      this.db = this.client.db(database);
      this.connected = true;
      this.connecting = false;
      console.log('Successfully connected to MongoDB');
    } catch (err) {
      console.error('MongoDB connection error:', err.message);
      this.connected = false;
      this.connecting = false;
      // Retry connection after 5 seconds
      setTimeout(() => this.connect(), 5000);
    }
  }

  isAlive() {
    return this.connected;
  }

  async nbUsers() {
    if (!this.connected) {
      await this.connect();
    }
    try {
      const usersCollection = this.db.collection('users');
      return await usersCollection.countDocuments();
    } catch (err) {
      console.error('Error counting users:', err.message);
      return 0;
    }
  }

  async nbFiles() {
    if (!this.connected) {
      await this.connect();
    }
    try {
      const filesCollection = this.db.collection('files');
      return await filesCollection.countDocuments();
    } catch (err) {
      console.error('Error counting files:', err.message);
      return 0;
    }
  }
}

const dbClient = new DBClient();
export default dbClient;
