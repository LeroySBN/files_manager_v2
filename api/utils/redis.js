// Redis utils
import { createClient } from 'redis';
import 'dotenv/config';

// const { promisify } = require('util');

const redisURL = process.env.REDIS_URL;

class RedisClient {
  constructor() {
    this.client = createClient({
      url: redisURL,
    });
    this.client.on('error', (err) => {
      console.log(`Redis client not connected to the server: ${err.message}`);
    });
    this.client.connect();
  }

  isAlive() {
    return this.client.isReady;
  }

  async get(key) {
    // const asyncGet = promisify(this.client.get).bind(this.client);
    try {
      // const value = await asyncGet(key);
      const value = await this.client.get(key);
      return value;
    } catch (error) {
      return null;
    }
  }

  async set(key, value, duration) {
    try {
      await this.client.setEx(key, duration, value);
    } catch (error) {
      console.log(error);
    }
  }

  async del(key) {
    try {
      await this.client.del(key);
    } catch (error) {
      console.log(error);
    }
  }
}

const redisClient = new RedisClient();

module.exports = redisClient;
