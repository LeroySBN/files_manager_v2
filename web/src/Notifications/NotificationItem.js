import React from "react";
import PropTypes from 'prop-types';
import { StyleSheet, css } from 'aphrodite';


class NotificationItem extends React.PureComponent {
  render() {
    const { type, value, html, markAsRead, id } = this.props;
    return (
      <React.Fragment>
        {type && value ? 
          <li
            className={
              type === "default" ? css(styles.default) : css(styles.urgent)
            }
            onClick={()=> markAsRead(id)}
            data-notification-type={type} >{value}
          </li>
        : null}
        {html ?
          <li
            className={css(styles.urgent)}
            onClick={()=> markAsRead(id)}
            data-notification-type="urgent"
            dangerouslySetInnerHTML={{__html: html}}
          />
        : null}
      </React.Fragment>
    );
  }
}

const styles = StyleSheet.create({
  default: {
    color: 'blue',
    cursor: 'pointer',
    '@media (max-width: 900px)': {
      borderBottom: '1px solid black',
      fontSize: '20px',
      padding: '10px 8px',
    },
  },
  urgent: {
    color: 'red',
    cursor: 'pointer',
    '@media (max-width: 900px)': {
      borderBottom: '1px solid black',
      fontSize: '20px',
      padding: '10px 8px',
    },
  },
});

NotificationItem.propTypes = {
  id: PropTypes.number,
  type: PropTypes.string.isRequired,
  value: PropTypes.string,
  __html: PropTypes.shape({ html: PropTypes.string }),
  markAsRead: PropTypes.func,
};

NotificationItem.defaultProps = {
  type: 'default',
  markAsRead: () => {},
  id: 0,
};

export default NotificationItem;
