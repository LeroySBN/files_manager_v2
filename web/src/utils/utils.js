export const getFullYear = () => new Date().getFullYear();

export const getFooterCopy = (isIndex) => ( isIndex ? 'File Push' : 'Files  by Leroy main dashboard');

export const getLatestNotification = () => {
  return ('<strong>Urgent requirement</strong> - complete by EOD');
}
