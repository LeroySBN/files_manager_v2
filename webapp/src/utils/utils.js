export const getFullYear = () => new Date().getFullYear();

export const getLatestNotification = () => {
  return ('<strong>Urgent requirement</strong> - complete by EOD');
}

// Basic email and password validation
// export const validateCredentials = (email, password) => {
//   const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
//   return emailRegex.test(email) && password.length >= 8;
// };
