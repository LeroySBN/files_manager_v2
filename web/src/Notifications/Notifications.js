import React from 'react';
import PropTypes from 'prop-types';
import { StyleSheet, css } from 'aphrodite';
import closeIcon from '../assets/close-icon.png';
import NotificationItem from './NotificationItem';
import { NotificationItemShape } from './NotificationItemShape';

class Notifications extends React.PureComponent {
  constructor(props) {
    super(props);
  }

  render() {
    return (
      <React.Fragment>
        {!this.props.displayDrawer ? (
          <div
            className={css(styles.menuItem)}
            onClick={()=> {
              this.props.handleDisplayDrawer();
            }}
          >
            <p>Your notifications</p>
          </div>
        ) : (
          <div className={css(styles.Notifications)}>
            <button 
              style={{ display: "flex", justifyContent: "flex-end", margin: "6px 6px 0 0", padding: "0", border: "none", cursor: "pointer", position: "absolute", top: "2px", right: "2px" }}
              aria-label="Close" 
              onClick={() => {
                this.props.handleHideDrawer();
              }}
            >
              <img className={css(styles.closeIcon)} src={closeIcon} alt='closeIcon' />
            </button>

            {this.props.listNotifications && this.props.listNotifications.length > 0 ? (
              <div>
                <p>Here is the list of notifications</p>
                <ul className={css(styles.ul)}>
                  {this.props.listNotifications.map(({ id, type, value, html }) => (
                    <NotificationItem
                      id={id}
                      key={id}
                      type={type}
                      value={value}
                      html={html}
                      markAsRead={this.props.markNotificationAsRead}
                    />
                  ))}
                </ul>
              </div>
            ) : (
            <p>No new notification for now</p>
            )}
          </div>
        )
        }
      </React.Fragment>
    );
  }
}

const opacityKeyframes = {
  'from': {
      opacity: 0.5,
  },

  'to': {
      opacity: 1,
  }
};

const translateKeyframes = {
  '0%': {
      transform: 'translateY(0)',
  },

  '50%': {
      transform: 'translateY(-5px)',
  },

  '100%': {
      transform: 'translateY(5)',
  },
};

const styles = StyleSheet.create({
  menuItem: {
    position: 'absolute',
    top: '0.5rem',
    right: '0.5rem',
    padding: '4px 16px',
    ':hover': {
      cursor: 'pointer',
      backgroundColor: '#fff8f8',
      animationName: [opacityKeyframes, translateKeyframes],
      animationDuration: '1s, 0.5s',
      animationIterationCount: '3',
    },
  },
  Notifications: {
    position: 'absolute',
    top: '1rem',
    right: '1em',
    padding: '8px 16px',
    border: '2px #e1003c dotted',
    color: '#000000',
    fontFamily: 'Helvetica, sans-serif',
    fontSize: '1rem',
    textAlign: 'left',
    backgroundColor: '#fff8f8',
    '@media (max-width: 900px)': {
      top: '0',
      right: '0',
      margin: '0',
      padding: '0',
      width: '100%',
      height: '100%',
      border: 'none',
      zIndex: '3',
    },
  },
  closeIcon: {
    margin: '0',
    height: '15px',
  },
  ul: {
    '@media (max-width: 900px)': {
      listStyle: 'none',
      padding: '0',
      fontSize: '20px',
    },
  }
});

Notifications.propTypes = {
  displayDrawer: PropTypes.bool,
  listNotifications: PropTypes.arrayOf(NotificationItemShape),
  handleDisplayDrawer: PropTypes.func,
  handleHideDrawer: PropTypes.func,
  markNotificationAsRead: PropTypes.func,
};

Notifications.defaultProps = {
  displayDrawer: false,
  listNotifications: [],
  handleDisplayDrawer: () => {},
  handleHideDrawer: () => {},
  markNotificationAsRead: ()=> {},
};

export default Notifications;
