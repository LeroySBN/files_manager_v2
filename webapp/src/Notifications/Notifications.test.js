import React from 'react';
import { shallow, mount } from 'enzyme';
import Notifications from './Notifications';
import NotificationItem from './NotificationItem';
import { StyleSheetTestUtils } from "aphrodite";
import { getLatestNotification } from '../utils/utils';

beforeEach(() => {
  StyleSheetTestUtils.suppressStyleInjection();
});

afterEach(() => {
  StyleSheetTestUtils.clearBufferAndResumeStyleInjection();
});

const listNotifications = [
  { id: 1, type: "default", value: "New course available" },
  { id: 2, type: "urgent", value: "New resume available" },
  { id: 3, type: "urgent", html: getLatestNotification() },
];


describe('Notifications component testing', () => {
  it('Notifications renders without crashing', () => {
    const component = shallow(<Notifications />);
    expect(component).toMatchSnapshot();
  });

  it('Notifications renders a list with three items', () => {
    const component = shallow(<Notifications />);
    expect(component.contains(<NotificationItem />)).exist;
  });

  it('Notifications renders the text "Here is the list of notifications"', () => {
    const component = shallow(<Notifications />);
    expect(component.contains(<p>Here is the list of notifications</p>)).exist
  });

  // it('menu item is being displayed when displayDrawer is false', () => {
  //   const component = shallow(<Notifications />);
  //   expect(component.find('.menuItem')).toHaveLength(1);
  // });

  it('div.Notifications is not being displayed when displayDrawer is false', () => {
    const component = shallow(<Notifications />);
    expect(component.find('.Notifications')).toHaveLength(0);
  });

  // it('menu item is being displayed when displayDrawer is true', () => {
  //   const component = shallow(<Notifications displayDrawer={true} />);
  //   expect(component.find('.menuItem')).toHaveLength(1);
  // });

  // it('div.Notifications is being displayed when displayDrawer is true', () => {
  //   const component = shallow(<Notifications displayDrawer={true} />);
  //   expect(component.find('.Notifications')).toHaveLength(1);
  // });
  
  // it('should call console.log with the right message when markAsRead is called', () => {
  //   const markNotificationAsRead = jest.fn();
  //   const wrapper = mount(
  //     <Notifications
  //       displayDrawer
  //       listNotifications={listNotifications}
  //       markNotificationAsRead={markNotificationAsRead}
  //     />
  //   );

  //   const instance = wrapper.instance();

  //   // Call the markAsRead function
  //   instance.props.markNotificationAsRead(1);

  //   // Verify that console.log was called with the expected message
  //   expect(console.log).toHaveBeenCalledWith('Notification 1 has been marked as read');

  //   // Restore the original console.log function
  //   console.log.mockRestore();
  // });

  it('should not rerender when updating with the same list', () => {
    const wrapper = mount(
      <React.StrictMode>
        <Notifications listNotifications={listNotifications} />
      </React.StrictMode>
    );
  
    const updateSpy = jest.spyOn(wrapper.instance(), 'forceUpdate');
    const updatedList = listNotifications.slice(); // Copy the list
  
    wrapper.setProps({ listNotifications: updatedList });
  
    expect(updateSpy).toHaveBeenCalledTimes(0);
  
    // Clean up the spy
    updateSpy.mockRestore();
  });
  

  it('should rerender when updating with a longer list', () => {
    const listNotifications = [
      { id: 1, type: 'default', value: 'Notification 1' },
      { id: 2, type: 'urgent', html: '<strong>Notification 2</strong>' },
    ];
  
    const wrapper = mount(<Notifications listNotifications={listNotifications} />);
    const component = wrapper.find('Notifications');
  
    const shouldRerender = jest.spyOn(component.instance(), 'render');
    const updatedList = [...listNotifications, { id: 3, type: 'default', value: 'Notification 3' }];
  
    wrapper.setProps({ listNotifications: updatedList });
  
    expect(shouldRerender).toHaveBeenCalledTimes(1);
  });

  it('verify that clicking on the menu item calls handleDisplayDrawer', () => {
    const handleDisplayDrawer = jest.fn();
    const wrapper = shallow(<Notifications handleDisplayDrawer={handleDisplayDrawer} />);
    const spy = jest.spyOn(wrapper.instance().props, 'handleDisplayDrawer');
    wrapper.find('div').first().simulate('click');
    expect(spy).toHaveBeenCalled();
    spy.mockRestore();
  });
  
});
