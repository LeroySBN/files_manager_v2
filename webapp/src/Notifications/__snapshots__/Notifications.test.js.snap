// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notifications component testing Notifications renders without crashing 1`] = `ShallowWrapper {}`;
